### PR TITLE
add nlohmann-json to rosdistro

### PR DESCRIPTION
**Pull request by hakuturu583**
I want to add nlohmann-json to rosdistro.

Fedra package is here : https://rpms.remirepo.net/rpmphp/zoom.php?rpm=json
Gentoo package is here : https://packages.gentoo.org/packages/dev-cpp/nlohmann_json
Ubuntu package is here : https://packages.ubuntu.com/search?keywords=nlohmann-json-dev
Debian package is here : https://packages.debian.org/stretch/nlohmann-json-dev

### CODE REVIEW

**Review by tfoote — APPROVED** [review]

File: rosdep/base.yaml [hunk 1]
```diff
@@ -4617,6 +4617,11 @@ nkf:
   fedora: [nkf]
   gentoo: [app-i18n/nkf]
   ubuntu: [nkf]
+nlohmann-json:
+  debian: [nlohmann-json-dev]
+  fedra: [json]
+  gentoo: [nlohmann_json]
+  ubuntu: [nlohmann-json-dev ]
```
Comment (clalancette) [inline rosdep/base.yaml:4624]: This looks to only be available in Bionic and later.  Since we still support Xenial, you'll need to break this rule out into a "null" for xenial, and the package name here for the wildcard.  See elsewhere in this file on how to do that.  Also, there is a trailing whitespace in the brackets (that I'm surprised our CI didn't catch).

File: rosdep/base.yaml [hunk 2]
```diff
@@ -4617,6 +4617,11 @@ nkf:
   fedora: [nkf]
   gentoo: [app-i18n/nkf]
   ubuntu: [nkf]
+nlohmann-json:
+  debian: [nlohmann-json-dev]
+  fedra: [json]
```
Comment (clalancette) [inline rosdep/base.yaml:4622]: This should be `fedora`.  Also, while the link you pasted *does* show this as the nlohmann version of json, I actually don't think that it is the same one.  If you look at https://apps.fedoraproject.org/packages/json , it seems to be a json-rpc package from glib.  Since I don't see another equivalent package for Fedora, let's just remove it here.

File: rosdep/base.yaml [hunk 3]
```diff
@@ -4617,6 +4617,11 @@ nkf:
...
+  gentoo: [nlohmann_json]
```
Comment (clalancette) [inline rosdep/base.yaml:4623]: I don't think this is the right format for Gentoo, but I'm not entirely sure.  Most of the other Gentoo keys seem to be of the sort `dev-cpp/nlohmann_json`.

File: rosdep/base.yaml [hunk 4]
```diff
@@ -4617,6 +4617,11 @@ nkf:
   fedora: [nkf]
   gentoo: [app-i18n/nkf]
   ubuntu: [nkf]
+nlohmann-json:
```
Comment (clalancette) [inline rosdep/base.yaml:4620]: Since you are going for the development packages, this key should be `nlohmann-json-dev`.  That way, we can add (either in this PR, or a later one) a separate key of `nlohmann-json` which is just the runtime portions.

**Review by clalancette — CHANGES_REQUESTED** [review]

File: rosdep/base.yaml (hunk 1 above)
Comment (tfoote) [reply rosdep/base.yaml:4624]: @clalancette historically we've let this slide. It's known not to break anyone who's using the older platform because they are guaranteed not to be using the rule currently and saves us complexity in our management. That was particularly important before we had the wildcard support. Now we do have the ability to do this without incurring the cost of repeatedly adding new rules so I guess it's worth doing now.

File: rosdep/base.yaml (hunk 2 above)
Comment (tfoote) [reply rosdep/base.yaml:4622]: good catch

File: rosdep/base.yaml (hunk 3 above)
Comment (tfoote) [reply rosdep/base.yaml:4623]: @allenh1

File: rosdep/base.yaml (hunk 4 above)
Comment (tfoote) [reply rosdep/base.yaml:4620]: +1

File: rosdep/base.yaml (hunk 3 above)
Comment (allenh1) [reply rosdep/base.yaml:4623]: > Most of the other Gentoo keys seem to be of the sort `dev-cpp/nlohmann_json`

@clalancette This is true. The reason behind this is that some packages have different categories but the same package name. For example, the ROS overlay uses `ros-[distro]/[package]` for the version of the package per ROS distribution (just to keep things organized).

*TL;DR*: this _would_ work, but I strongly prefer it be `dev-cpp/nlohmann_json` instead of just the bare `nlohmann_json`.

@hakuturu583 thanks for including the Gentoo keys in this patch! I very much appreciate it.

File: rosdep/base.yaml (hunk 1 above)
Comment (hakuturu583) [reply rosdep/base.yaml]: I fixed this issue in https://github.com/ros/rosdistro/pull/24432/commits/000e1763d563544e2eec3f54a6bba064e2aac610

File: rosdep/base.yaml (hunk 2 above)
Comment (hakuturu583) [reply rosdep/base.yaml]: Thanks! I fixed in https://github.com/ros/rosdistro/pull/24432/commits/600a51b70b054f97ce633e5a1839a427e504a094

File: rosdep/base.yaml (hunk 3 above)
Comment (hakuturu583) [reply rosdep/base.yaml]: Thanks! I fixed this issue in https://github.com/ros/rosdistro/pull/24432/commits/925e15528003f84a2c05d9cb8cebf59c1a5be4df

File: rosdep/base.yaml (hunk 4 above)
Comment (hakuturu583) [reply rosdep/base.yaml]: Thanks! I fixed in https://github.com/ros/rosdistro/pull/24432/commits/8178b1406ac6142cf17d622437727cdb1a75fe6f

**Comment by hakuturu583** [comment]
Hi, I found my PR was failed in running Nosetest.
And I found the error is below.

```

**Comment by hakuturu583** [comment]
What is the error and what is the nosetests

**Comment by clalancette** [comment]
> Hi, I found my PR was failed in running Nosetest.

So nosetest is a standard way of running python tests: https://nose.readthedocs.io/en/latest/ .  It's what we use to run our unittests.

Usually there is a line in the test that points out exactly which line failed, but for some reason it is not showing it here.  Running it locally, I see that the problem is that there is a trailing whitespace after the `ubuntu: ` line.  There is also still trailing whitespace after nlohmann-json-dev .  Both of those should be fixed, then I think CI will be happier and I'll be happy to approve.

**Comment by hakuturu583** [comment]
@clalancette Thanks! I removed whitespace.

**Review by clalancette — APPROVED** [review]
